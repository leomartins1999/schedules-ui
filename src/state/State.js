export function SuccessState(value) {
    return { status: "SUCCESS", value }
}

export function LoadingState() {
    return { status: "LOADING" }
}

export function ErrorState() {
    return { status: "ERROR" }
}
