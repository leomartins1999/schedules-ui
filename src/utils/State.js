import Spinner from "react-bootstrap/Spinner";
import Alert from "react-bootstrap/Alert";

export const SUCCESS_STATE = "SUCCESS"
export const LOADING_STATE = "LOADING"
export const ERROR_STATE = "ERROR"

const ERROR_MESSAGE = "An error has occurred!"

export function SuccessState(value) {
    return { status: SUCCESS_STATE, value }
}

export function LoadingState() {
    return { status: LOADING_STATE }
}

export function ErrorState() {
    return { status: ERROR_STATE }
}

export function renderStatefulContent(statefulContent, onSuccess) {
    switch (statefulContent.status) {
        case ERROR_STATE: return <Alert variant="danger">{ERROR_MESSAGE}</Alert>;
        case LOADING_STATE: return <Spinner animation="border" />
        default: return onSuccess(statefulContent.value);
    }
}
