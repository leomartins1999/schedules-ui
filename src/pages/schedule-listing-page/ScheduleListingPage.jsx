import { useEffect, useState } from "react";
import ListGroup from "react-bootstrap/ListGroup";
import Spinner from "react-bootstrap/Spinner";
import Alert from "react-bootstrap/Alert";

import { ScheduleListItem, CreateScheduleForm } from "../../components/index";
import { LoadingState } from "../../utils/State";

function renderSchedules(schedulesState) {
  switch (schedulesState.status) {
    case "LOADING": return <Spinner animation="border" />;
    case "ERROR": return <Alert variant="danger">An error has occured loading the schedules</Alert>;
    case "SUCCESS": return schedulesState.value.map(s => <ScheduleListItem key={s.id} schedule={s} />);
  }
}

function ScheduleListingPage({ service }) {
  const [schedulesState, setSchedulesState] = useState(LoadingState());

  useEffect(async () => {
    const result = await service.getSchedules();
    setSchedulesState(result);
  }, []);

  return (
    <>
      <ListGroup>
        <h2>Schedules</h2>
        {renderSchedules(schedulesState)}
      </ListGroup>
      <CreateScheduleForm service={service} />
    </>
  );
}

export default ScheduleListingPage;
