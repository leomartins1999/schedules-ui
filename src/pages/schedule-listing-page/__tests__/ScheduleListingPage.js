import renderer from 'react-test-renderer';
import ScheduleListingPage from '../ScheduleListingPage';
import { ErrorState, LoadingState, SuccessState } from '../../../state/State';

it('renders no schedules', () => {
    const service = {
        getSchedules: async () => SuccessState([])
    }

    const tree = renderer
        .create(<ScheduleListingPage service={service} />)
        .toJSON()

    expect(tree).toMatchSnapshot();
})

it('renders the schedules list', () => {
    const service = {
        getSchedules: async () => SuccessState([
            {
                id: "1",
                name: "Done schedule",
                status: "DONE"
            },
            {
                id: "2",
                name: "Importing schedule",
                status: "IMPORTING"
            }
        ])
    }

    const tree = renderer
        .create(<ScheduleListingPage service={service} />)
        .toJSON()

    expect(tree).toMatchSnapshot();
})

it('renders loading', () => {
    const service = {
        getSchedules: async () => LoadingState()
    }

    const tree = renderer
        .create(<ScheduleListingPage service={service} />)
        .toJSON()

    expect(tree).toMatchSnapshot();
})

it('renders error', () => {
    const service = {
        getSchedules: async () => ErrorState()
    }

    const tree = renderer
        .create(<ScheduleListingPage service={service} />)
        .toJSON()

    expect(tree).toMatchSnapshot();
})
