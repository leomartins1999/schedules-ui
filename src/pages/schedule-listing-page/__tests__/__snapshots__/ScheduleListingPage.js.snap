// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders error 1`] = `
Array [
  <div
    className="list-group"
    onKeyDown={[Function]}
  >
    <h2>
      Schedules
    </h2>
    <div
      className="spinner-border"
    />
  </div>,
  <div
    className="pt-2 pb-2 accordion"
  >
    <div
      className="accordion-item"
    >
      <h2
        className="accordion-header"
      >
        <button
          aria-expanded={false}
          className="accordion-button collapsed"
          onClick={[Function]}
          type="button"
        >
          Upload new Schedule
        </button>
      </h2>
      <div
        aria-expanded={null}
        className="accordion-collapse collapse"
      >
        <div
          className="accordion-body"
        >
          <form
            className="pt-1"
            onSubmit={[Function]}
          >
            <div
              className="mb-3"
            >
              <label
                className="form-label"
                htmlFor="formBasicEmail"
              >
                Schedule name
              </label>
              <input
                className="form-control"
                id="formBasicEmail"
                onChange={[Function]}
                placeholder="Enter schedule name"
                required={true}
                type="text"
              />
            </div>
            <div
              className="mb-3"
            >
              <label
                className="form-label"
                htmlFor="formFile"
              >
                File
              </label>
              <input
                className="form-control"
                id="formFile"
                onChange={[Function]}
                required={true}
                type="file"
              />
            </div>
            <button
              className="btn btn-primary"
              disabled={false}
              type="submit"
            >
              Create Schedule
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`renders loading 1`] = `
Array [
  <div
    className="list-group"
    onKeyDown={[Function]}
  >
    <h2>
      Schedules
    </h2>
    <div
      className="spinner-border"
    />
  </div>,
  <div
    className="pt-2 pb-2 accordion"
  >
    <div
      className="accordion-item"
    >
      <h2
        className="accordion-header"
      >
        <button
          aria-expanded={false}
          className="accordion-button collapsed"
          onClick={[Function]}
          type="button"
        >
          Upload new Schedule
        </button>
      </h2>
      <div
        aria-expanded={null}
        className="accordion-collapse collapse"
      >
        <div
          className="accordion-body"
        >
          <form
            className="pt-1"
            onSubmit={[Function]}
          >
            <div
              className="mb-3"
            >
              <label
                className="form-label"
                htmlFor="formBasicEmail"
              >
                Schedule name
              </label>
              <input
                className="form-control"
                id="formBasicEmail"
                onChange={[Function]}
                placeholder="Enter schedule name"
                required={true}
                type="text"
              />
            </div>
            <div
              className="mb-3"
            >
              <label
                className="form-label"
                htmlFor="formFile"
              >
                File
              </label>
              <input
                className="form-control"
                id="formFile"
                onChange={[Function]}
                required={true}
                type="file"
              />
            </div>
            <button
              className="btn btn-primary"
              disabled={false}
              type="submit"
            >
              Create Schedule
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`renders no schedules 1`] = `
Array [
  <div
    className="list-group"
    onKeyDown={[Function]}
  >
    <h2>
      Schedules
    </h2>
    <div
      className="spinner-border"
    />
  </div>,
  <div
    className="pt-2 pb-2 accordion"
  >
    <div
      className="accordion-item"
    >
      <h2
        className="accordion-header"
      >
        <button
          aria-expanded={false}
          className="accordion-button collapsed"
          onClick={[Function]}
          type="button"
        >
          Upload new Schedule
        </button>
      </h2>
      <div
        aria-expanded={null}
        className="accordion-collapse collapse"
      >
        <div
          className="accordion-body"
        >
          <form
            className="pt-1"
            onSubmit={[Function]}
          >
            <div
              className="mb-3"
            >
              <label
                className="form-label"
                htmlFor="formBasicEmail"
              >
                Schedule name
              </label>
              <input
                className="form-control"
                id="formBasicEmail"
                onChange={[Function]}
                placeholder="Enter schedule name"
                required={true}
                type="text"
              />
            </div>
            <div
              className="mb-3"
            >
              <label
                className="form-label"
                htmlFor="formFile"
              >
                File
              </label>
              <input
                className="form-control"
                id="formFile"
                onChange={[Function]}
                required={true}
                type="file"
              />
            </div>
            <button
              className="btn btn-primary"
              disabled={false}
              type="submit"
            >
              Create Schedule
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`renders the schedules list 1`] = `
Array [
  <div
    className="list-group"
    onKeyDown={[Function]}
  >
    <h2>
      Schedules
    </h2>
    <div
      className="spinner-border"
    />
  </div>,
  <div
    className="pt-2 pb-2 accordion"
  >
    <div
      className="accordion-item"
    >
      <h2
        className="accordion-header"
      >
        <button
          aria-expanded={false}
          className="accordion-button collapsed"
          onClick={[Function]}
          type="button"
        >
          Upload new Schedule
        </button>
      </h2>
      <div
        aria-expanded={null}
        className="accordion-collapse collapse"
      >
        <div
          className="accordion-body"
        >
          <form
            className="pt-1"
            onSubmit={[Function]}
          >
            <div
              className="mb-3"
            >
              <label
                className="form-label"
                htmlFor="formBasicEmail"
              >
                Schedule name
              </label>
              <input
                className="form-control"
                id="formBasicEmail"
                onChange={[Function]}
                placeholder="Enter schedule name"
                required={true}
                type="text"
              />
            </div>
            <div
              className="mb-3"
            >
              <label
                className="form-label"
                htmlFor="formFile"
              >
                File
              </label>
              <input
                className="form-control"
                id="formFile"
                onChange={[Function]}
                required={true}
                type="file"
              />
            </div>
            <button
              className="btn btn-primary"
              disabled={false}
              type="submit"
            >
              Create Schedule
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>,
]
`;
